### PR TITLE
[TFL] Support I32 for OptimizeSlice pattern

### PR DESCRIPTION
**Pull request by WindQAQ**
Somehow, some int32 `begin` will be generated. This PR extends the original pattern to support i32 `begin` because we have type constraint of `TFL_I32OrI64Tensor` on `begin` in ODS. Pattern is found in `tf.keras.layers.MultiHeadAttention`, for example.

```python
import tensorflow as tf

layer = tf.keras.layers.MultiHeadAttention(num_heads=2, key_dim=2)
target = tf.keras.Input(shape=[8, 16], batch_size=1)
source = tf.keras.Input(shape=[4, 16], batch_size=1)
output_tensor = layer(target, source, return_attention_scores=False)
model = tf.keras.Model([target, source], output_tensor)
```

Also fix:
- when input has dynamic shape.
- when shape has -1.

It turns out that we have to check if `shape` is the shape of `input`. Otherwise, given `input: tensor<?xf32>` and `output: tensor<?xf32>`, the original pattern is not going to work.

### CODE REVIEW

File: tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td [hunk 1]
```diff
@@ -760,17 +760,15 @@ foreach ArgMinMaxOp = [TFL_ArgMinOp, TFL_ArgMaxOp] in {
...
+  "($0.isa<DenseIntElementsAttr>() && "
+  "llvm::all_of($0.cast<DenseIntElementsAttr>().getIntValues(), "
+               "[](APInt v){ return v.getSExtValue() == static_cast<int64_t>(" #val# ");}))">>;
 
 // Remove Slice ops slicing the whole input tensor, effectively no-op
 def OptimizeSliceOp : Pat<
   (TFL_SliceOp:$output $input, (ConstantOp $begin), $shape),
   (replaceWithValue $input),
-  [(AllElementsAreI64<"0"> $begin),
+  [(AllElementsAreInt<"0"> $begin),
```
Comment (abattery) [inline tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td:772]: can we limit only int32 and int64 inputs?

Comment (WindQAQ) [reply tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td:772]: Of course we can, but we have already had type constraint in ODS so begin will be either i32 or i64, not sure if it will be a redundant check?

Comment (abattery) [reply tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td:772]: This optimization pattern will be applied after TF -> TFL legalization. It is better to check the types in here.

Comment (WindQAQ) [reply tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td]: Got it! Updated 😃

**Review by abattery — APPROVED** [review]

**Comment by abattery** [comment]
There is an error due after this new pattern. Please take a look.

[Before]
module attributes {tf.versions = {bad_consumers = [], min_consumer = 12 : i32, producer = 685 : i32}}  {
  func @main(%arg0: tensor<f32> {tf.device = "/device:CPU:0"}) -> tensor<?xf32> attributes {tf.entry_function = {control_outputs = "", inputs = "Placeholder", outputs = "TensorListStack"}} {
    %cst = constant dense<0> : tensor<i32>
    %cst_0 = constant dense<0.000000e+00> : tensor<4xf32>
    %0:3 = "tfl.while"(%cst, %cst, %cst_0) ( {
    ^bb0(%arg1: tensor<i32>, %arg2: tensor<i32>, %arg3: tensor<?xf32>):  // no predecessors
      %cst_1 = constant dense<4> : tensor<i32>
      %1 = "tfl.less"(%arg2, %cst_1) : (tensor<i32>, tensor<i32>) -> tensor<i1>
      "tfl.yield"(%1) : (tensor<i1>) -> ()
    },  {
    ^bb0(%arg1: tensor<i32>, %arg2: tensor<i32>, %arg3: tensor<?xf32>):  // no predecessors
      %cst_1 = constant dense<1.000000e+00> : tensor<f32>
      %cst_2 = constant dense<1> : tensor<i32>
      %cst_3 = constant dense<0> : tensor<1xi32>
      %cst_4 = constant dense<-1> : tensor<1xi32>
      %cst_5 = constant dense<1> : tensor<1xi32>
      %1 = tfl.add %arg0, %arg0 {fused_activation_function = "NONE"} : tensor<f32>
      %2 = tfl.add %1, %cst_1 {fused_activation_function = "NONE"} : tensor<f32>
      %3 = tfl.add %arg2, %cst_2 {fused_activation_function = "NONE"} : tensor<i32>
      %4 = "tfl.reshape"(%arg2, %cst_5) : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
      %5 = "tfl.slice"(%arg3, %cst_3, %4) : (tensor<?xf32>, tensor<1xi32>, tensor<1xi32>) -> tensor<?xf32>
      %6 = "tfl.reshape"(%3, %cst_5) : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
      %7 = "tfl.slice"(%arg3, %6, %cst_4) : (tensor<?xf32>, tensor<1xi32>, tensor<1xi32>) -> tensor<?xf32>
      %8 = "tfl.reshape"(%2, %cst_5) : (tensor<f32>, tensor<1xi32>) -> tensor<1xf32>
      %9 = "tfl.concatenation"(%5, %8, %7) {axis = 0 : i32, fused_activation_function = "NONE"} : (tensor<?xf32>, tensor<1xf32>, tensor<?xf32>) -> tensor<?xf32>
      %10 = tfl.add %arg1, %cst_2 {fused_activation_function = "NONE"} : tensor<i32>
      "tfl.yield"(%10, %3, %9) : (tensor<i32>, tensor<i32>, tensor<?xf32>) -> ()
    }) {is_stateless = true} : (tensor<i32>, tensor<i32>, tensor<4xf32>) -> (tensor<i32>, tensor<i32>, tensor<?xf32>)
    return %0#2 : tensor<?xf32>
  }
}

[After]
module attributes {tf.versions = {bad_consumers = [], min_consumer = 12 : i32, producer = 685 : i32}}  {
  func @main(%arg0: tensor<f32> {tf.device = "/device:CPU:0"}) -> tensor<?xf32> attributes {tf.entry_function = {control_outputs = "", inputs = "Placeholder", outputs = "TensorListStack"}} {
    %cst = constant dense<0> : tensor<i32>
    %cst_0 = constant dense<0.000000e+00> : tensor<4xf32>
    %0:3 = "tfl.while"(%cst, %cst, %cst_0) ( {
    ^bb0(%arg1: tensor<i32>, %arg2: tensor<i32>, %arg3: tensor<?xf32>):  // no predecessors
      %cst_1 = constant dense<4> : tensor<i32>
      %1 = "tfl.less"(%arg2, %cst_1) : (tensor<i32>, tensor<i32>) -> tensor<i1>
      "tfl.yield"(%1) : (tensor<i1>) -> ()
    },  {
    ^bb0(%arg1: tensor<i32>, %arg2: tensor<i32>, %arg3: tensor<?xf32>):  // no predecessors
      %cst_1 = constant dense<1.000000e+00> : tensor<f32>
      %cst_2 = constant dense<1> : tensor<i32>
      %cst_3 = constant dense<-1> : tensor<1xi32>
      %cst_4 = constant dense<1> : tensor<1xi32>
      %1 = tfl.add %arg0, %arg0 {fused_activation_function = "NONE"} : tensor<f32>
      %2 = tfl.add %1, %cst_1 {fused_activation_function = "NONE"} : tensor<f32>
      %3 = tfl.add %arg2, %cst_2 {fused_activation_function = "NONE"} : tensor<i32>
      %4 = "tfl.reshape"(%3, %cst_4) : (tensor<i32>, tensor<1xi32>) -> tensor<1xi32>
      %5 = "tfl.slice"(%arg3, %4, %cst_3) : (tensor<?xf32>, tensor<1xi32>, tensor<1xi32>) -> tensor<?xf32>
      %6 = "tfl.reshape"(%2, %cst_4) : (tensor<f32>, tensor<1xi32>) -> tensor<1xf32>
      %7 = "tfl.concatenation"(%arg3, %6, %5) {axis = 0 : i32, fused_activation_function = "NONE"} : (tensor<?xf32>, tensor<1xf32>, tensor<?xf32>) -> tensor<?xf32>
      %8 = tfl.add %arg1, %cst_2 {fused_activation_function = "NONE"} : tensor<i32>
      "tfl.yield"(%8, %3, %7) : (tensor<i32>, tensor<i32>, tensor<?xf32>) -> ()
    }) {is_stateless = true} : (tensor<i32>, tensor<i32>, tensor<4xf32>) -> (tensor<i32>, tensor<i32>, tensor<?xf32>)
    return %0#2 : tensor<?xf32>
  }
}

**Comment by abattery** [comment]
To correct the above case, the shape argument should be checked to verify that it covers the original shape.

File: tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td [hunk 2]
```diff
@@ -760,17 +760,17 @@ foreach ArgMinMaxOp = [TFL_ArgMinOp, TFL_ArgMaxOp] in {
...
+class AllElementsAreI32OrI64<string val> : Constraint<CPred<
+  "($0.isa<DenseIntElementsAttr>() && "
+  "($0.cast<DenseIntElementsAttr>().getType().cast<ShapedType>().getElementType().isInteger(32) || "
+  "$0.cast<DenseIntElementsAttr>().getType().cast<ShapedType>().getElementType().isInteger(64)) && "
+  "llvm::all_of($0.cast<DenseIntElementsAttr>().getIntValues(), "
+               "[](APInt v){ return v.getSExtValue() == static_cast<int64_t>(" #val# ");}))">>;
 
 // Remove Slice ops slicing the whole input tensor, effectively no-op
 def OptimizeSliceOp : Pat<
   (TFL_SliceOp:$output $input, (ConstantOp $begin), $shape),
```
Comment (abattery) [inline tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td:772]: To correct the above case, the shape argument should be checked to verify that it covers the original shape.

**Comment by WindQAQ** [comment]
Looks like `IsTailOfShape` will allow dynamic dim, which should not happen in this pattern.

File: tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td (hunk 2 above)
Comment (abattery) [reply tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td:772]: Could you also add more test cases to cover the above case?

Comment (WindQAQ) [reply tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td:772]: Sure, I'll update it as soon as possible.

Comment (WindQAQ) [reply tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td]: It should work now. We are now checking shape of `input` with `shape`. This requires `shape` to be a constant. Let me know if it works smoothly :smiley:

File: tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td [hunk 3]
```diff
@@ -760,17 +760,28 @@ foreach ArgMinMaxOp = [TFL_ArgMinOp, TFL_ArgMaxOp] in {
...
 
-class AllElementsAreI64<string val> : Constraint<CPred<
-  "($0.isa<DenseElementsAttr>() && "
-   "$0.cast<DenseElementsAttr>().getType().cast<ShapedType>().getElementType().isInteger(64) && "
-   "std::all_of($0.cast<DenseElementsAttr>().getValues<int64_t>().begin(), "
-               "$0.cast<DenseElementsAttr>().getValues<int64_t>().end(), "
-               "[](int64_t v){ return v == " #val# ";}))">>;
-
-// Remove Slice ops slicing the whole input tensor, effectively no-op
+class AllElementsAreI32OrI64<string val> : Constraint<CPred<
```
Comment (abattery) [inline tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td:763]: AllElementsAreI32OrI64 and IsSliceWholeInput are complicated. It would be better to move the logic under the optimize.cc for the readability.

**Comment by WindQAQ** [comment]
@abattery Sorry for the delay. See if this is more readable. I follow the naming convention of `CanOptimizeIdentityGatherNdOrScatterNdOp` in the same file.

**Review by abattery — APPROVED** [review]

File: tensorflow/compiler/mlir/lite/transforms/optimize.cc [hunk 4]
```diff
@@ -210,6 +210,50 @@ bool CanOptimizeIdentityGatherNdOrScatterNdOp(Value params,
...
+  if (rank != begin_attr.getNumElements() ||
+      rank != size_attr.getNumElements()) {
+    return false;
+  }
+
+  // Checks if `begin` is all 0s, and `size[i]` is equal to either -1 or
+  // `input.shape[i]`.
+  for (int64_t i = 0; i < rank; ++i) {
+    if (begin_attr.getValue<APInt>({i}).getSExtValue() != 0) return false;
+    int64_t si = size_attr.getValue<APInt>({i}).getSExtValue();
```
Comment (abattery) [inline tensorflow/compiler/mlir/lite/transforms/optimize.cc:250]: static_cast<unsigned long>(i)

File: tensorflow/compiler/mlir/lite/transforms/optimize.cc [hunk 5]
```diff
@@ -210,6 +210,50 @@ bool CanOptimizeIdentityGatherNdOrScatterNdOp(Value params,
...
+  int64_t rank = input_ty.getRank();
+  if (rank != begin_attr.getNumElements() ||
+      rank != size_attr.getNumElements()) {
+    return false;
+  }
+
+  // Checks if `begin` is all 0s, and `size[i]` is equal to either -1 or
+  // `input.shape[i]`.
+  for (int64_t i = 0; i < rank; ++i) {
+    if (begin_attr.getValue<APInt>({i}).getSExtValue() != 0) return false;
```
Comment (abattery) [inline tensorflow/compiler/mlir/lite/transforms/optimize.cc:249]: static_cast<unsigned long>(i)

File: tensorflow/compiler/mlir/lite/transforms/optimize.cc (hunk 4 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/lite/transforms/optimize.cc:250]: How about `for (uint64_t i = 0....)`?

Comment (abattery) [reply tensorflow/compiler/mlir/lite/transforms/optimize.cc:250]: SG

**Review by abattery — APPROVED** [review]
